### PR TITLE
Add merge API to CLI to have multiple sources for one sink

### PR DESCRIPTION
**Pull request by shadaj**
Add merge API to CLI to have multiple sources for one sink

### CODE REVIEW

**Comment by shadaj** [comment]
To be honest, I'm not quite happy with the abstractions here (`merge` is a magic attribute on Hydroflow ports), but I'd like to get this merged and then iterate on the internals just to unblock David.

File: hydro_cli/src/core/hydroflow_crate/mod.rs [hunk 1]
```diff
@@ -75,6 +75,8 @@ impl HydroflowCrate {
         sink: &mut dyn HydroflowSink,
     ) -> Result<()> {
         if let Ok(instantiated) = sink.instantiate(&self.on) {
+            // TODO(shadaj): if already in this map, we want to broadcast
+            assert!(!self.port_to_server.contains_key(&my_port));
```
Comment (MingweiSamuel) [inline hydro_cli/src/core/hydroflow_crate/mod.rs:79]: broadcast?

File: hydroflow_cli_integration/src/lib.rs [hunk 2]
```diff
@@ -393,3 +448,44 @@ where
...
+                }
+                Poll::Ready(None) => {
+                    // this happens infrequently, so OK to be O(n)
+                    sources.remove(i);
+                }
+                Poll::Pending => {
+                    i += 1;
+                }
+            }
+        }
```
Comment (MingweiSamuel) [inline hydroflow_cli_integration/src/lib.rs:481]: `.retain_mut()` might be a cleaner way to represent this. though maybe slower

File: hydroflow_cli_integration/src/lib.rs [hunk 3]
```diff
@@ -393,3 +448,44 @@ where
         self.sink.take().unwrap()
     }
 }
+
+#[pin_project]
+pub struct MergeSource<T, S: Stream<Item = T> + Send + Sync + ?Sized> {
+    marker: PhantomData<T>,
+    #[pin]
+    sources: Vec<Pin<Box<S>>>,
+}
```
Comment (MingweiSamuel) [inline hydroflow_cli_integration/src/lib.rs:457]: `pin_project` seems unnecessary here? Already have `Pin<Box<S>>`

Comment (MingweiSamuel) [reply hydroflow_cli_integration/src/lib.rs:457]: Worth noting there is also `StreamExt::merge()` https://docs.rs/tokio-stream/latest/tokio_stream/trait.StreamExt.html#method.merge and `StreamMap` https://docs.rs/tokio-stream/latest/tokio_stream/struct.StreamMap.html in tokio

**Review by MingweiSamuel — APPROVED** [review]

File: hydro_cli/src/core/hydroflow_crate/mod.rs (hunk 1 above)
Comment (shadaj) [reply hydro_cli/src/core/hydroflow_crate/mod.rs:79]: "tee"

File: hydroflow_cli_integration/src/lib.rs (hunk 2 above)
Comment (shadaj) [reply hydroflow_cli_integration/src/lib.rs:481]: yeah I wanted to short circuit when you get a ready, which is why I implemented it this way

**Pull request merged**